### PR TITLE
ISPN-10135 Performance Regression for HotRod-TX

### PR DESCRIPTION
**Pull request by pruivo**
* try to prepare the transaction in the primary owner

https://issues.redhat.com/browse/ISPN-10135

### CODE REVIEW

**Comment by pruivo** [comment]
run performance tests please
cs-hotrod

**Comment by gustavolira** [comment]
run performance tests please
cs-hotrod

**Comment by pruivo** [comment]
run performance tests please
cs-hotrod-tx

**Comment by gustavolira** [comment]
run performance tests please
cs-hotrod-tx

**Comment by danberindei** [comment]
@gustavolira the results page is empty

**Comment by gustavolira** [comment]
@danberindei we are looking into it

**Comment by danberindei** [comment]
@pruivo funny, this PR made me look at the [library regression test](http://perfrepo.mw.lab.eng.bos.redhat.com/reports/metric/3160), and there's a huge drop in library repl read performance as well

**Comment by gustavolira** [comment]
run performance tests please
cs-hotrod-tx

**Comment by gustavolira** [comment]
> @pruivo funny, this PR made me look at the [library regression test](http://perfrepo.mw.lab.eng.bos.redhat.com/reports/metric/3160), and there's a huge drop in library repl read performance as well

@danberindei it's because https://issues.jboss.org/browse/ISPN-11550

**Comment by diegolovison** [comment]
I deleted outdated/inconsistent report request/results.
I am going to rerun the tests because HotRod-TX should be stable.
Stable in terms of having a closer throughput and service time.

**Comment by diegolovison** [comment]
run performance tests please
cs-hotrod-tx

**Comment by diegolovison** [comment]
run performance tests please
cs-hotrod-tx

**Comment by diegolovison** [comment]
Looks good to me!

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transaction/operations/PrepareTransactionOperation.java [hunk 1]
```diff
@@ -80,8 +82,18 @@ protected void executeOperation(Channel channel) {
       channel.writeAndFlush(buf);
    }
 
+   @Override
+   protected void fetchChannelAndInvoke(int retryCount, Set<SocketAddress> failedServers) {
+      if (retryCount < modifications.size()) {
```
Comment (danberindei) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transaction/operations/PrepareTransactionOperation.java:87]: I'd simplify this to only use the key owner on the first try.

IMO retries should be rare enough that we shouldn't worry about the server we're retrying on being a key owner, and we're going to send the request to an unrelated server anyway if the 2nd modification's key is owned by the failed server.

Comment (pruivo) [reply client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transaction/operations/PrepareTransactionOperation.java:87]: I don't see any advantage in doing that. 
If there is a chance of hitting a primary owner of some key, I'll take it. 
If you are worried about `modifications.size()`, the list is an `ArrayList`; it just reads a field.

Comment (danberindei) [reply client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transaction/operations/PrepareTransactionOperation.java:87]: I wasn't worried about `size()`, I was just thinking it's simpler if we use the same logic here as we use for other multi-key operations: 

https://github.com/infinispan/infinispan/blob/a2a183a3e23b37eb9820c7c18bec211ef2dc2deb/client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/PutAllOperation.java#L76

Comment (pruivo) [reply client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/transaction/operations/PrepareTransactionOperation.java]: @danberindei  done.

**Comment by danberindei** [comment]
Merged, thanks Pedro!

**Pull request merged**